src/Http/MiddlewareStack.php

```diff
@@ -0,0 +1,151 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http;
+
+use Countable;
+
+/**
+ * Provides methods for creating and manipulating a 'stack' of
+ * middleware callables. This stack is used to process a request and response
+ * via \Cake\Http\Runner.
+ */
+class MiddlewareStack implements Countable
+{
+    /**
+     * The stack of middleware callables.
+     *
+     * @var array
+     */
+    protected $stack = [];
+
+    /**
+     * Get the middleware object at the provided index.
+     *
+     * @param int $index The index to fetch.
+     * @return callable|null Either the callable middleware or null
+     *   if the index is undefined.
+     */
+    public function get($index)
+    {
+        if (isset($this->stack[$index])) {
+            return $this->stack[$index];
+        }
+        return null;
```

Comment:
I'll add a test for this. I'd rather not raise an exception just yet as it will complicate the request 'Runner' which is the next pull request for this library.